File: backend/services/process_labels.py
"""
Etiquetas dos processos — Lote 5, Secção B, ponto 15.

O PEDIDO
  Marcadores personalizados ("Sub 35", "VIP", "Urgente") para segmentar
  processos, porque as fases já não chegam.

O QUE JÁ EXISTIA
  `labels: Optional[List[str]]` está no modelo desde sempre, persiste em
  `process_update` e vem nas projecções. Faltava o editor (o comentário
  no `ProcessDetails` prometia um Dialog que nunca foi construído) e
  faltava a FILTRAGEM: nem `build_process_list_query` nem
  `build_kanban_query` conheciam o campo.

PORQUE É QUE CONTINUAM A SER STRINGS
  Passar a objectos obrigaria a migrar os dados existentes e a mexer em
  todas as projecções, e a única coisa que os objectos dariam a mais era
  a cor — que se deriva do próprio texto, deterministicamente, no
  frontend (`utils/processLabels.js`). "VIP" fica da mesma cor em todo o
  lado sem uma segunda colecção para manter em dia.

NORMALIZAR À ESCRITA, NUNCA À LEITURA
  "VIP", "vip" e " VIP " são a mesma etiqueta para quem segmenta e três
  para o Mongo. Com a normalização na leitura, cada filtro teria de a
  repetir — e o primeiro que se esquecesse devolvia uma lista a menos
  sem ninguém reparar.

O CATÁLOGO TEM ÂMBITO DE TENANT, E ISSO NÃO É OPCIONAL
  Para filtrar é preciso saber que etiquetas existem, e a fonte honesta
  é o que está nos processos. Um `distinct` sem a condição de rede seria
  uma fuga nova, da mesma família das do Lote 4/5: o nome de uma
  campanha da concorrência no dropdown de quem não a devia ver.
"""
import logging
from typing import Any, Iterable, Optional

from database import db
from services.tenant_network import build_tenant_condition

logger = logging.getLogger(__name__)

#: Uma etiqueta é um marcador, não um campo de notas.
MAX_COMPRIMENTO = 40
#: Acima disto deixa de segmentar e passa a ser ruído no cabeçalho.
MAX_ETIQUETAS = 20

CAMPO = "labels"


def normalizar_etiquetas(valor: Any) -> list[str]:
    """Etiquetas limpas, sem repetições e em número comportável.

    Aceita lista ou string separada por vírgulas (é assim que chegam de
    um campo de texto). Compara sem olhar à capitalização mas **preserva
    a forma da primeira ocorrência**: quem escreve "Sub 35" não quer ver
    "sub 35" no crachá.

    Devolve sempre uma lista — uma lista vazia é uma resposta legítima
    (limpar as etiquetas de um processo), não um erro.
    """
    if valor is None:
        return []

    if isinstance(valor, str):
        cruas: Iterable[Any] = valor.split(",")
    elif isinstance(valor, (list, tuple, set)):
        cruas = valor
    else:
        return []

    etiquetas: list[str] = []
    vistas: set[str] = set()
    for crua in cruas:
        if crua is None:
            continue
        texto = str(crua).strip()[:MAX_COMPRIMENTO].strip()
        if not texto:
            continue
        chave = texto.casefold()
        if chave in vistas:
            continue
        vistas.add(chave)
        etiquetas.append(texto)
        if len(etiquetas) >= MAX_ETIQUETAS:
            break
    return etiquetas


def build_labels_condition(
    etiquetas: Any,
    logica: Optional[str] = "OR",
) -> Optional[dict]:
    """Condição Mongo para filtrar por etiquetas, ou ``None``.

    ``None`` quer dizer **não filtrar** — o chamador tem de o testar
    antes de a juntar à query. Um filtro sempre presente esconderia os
    processos sem etiqueta nenhuma, que são a maioria.

    ``logica`` espelha o ``assigned_logic`` que já existe neste ficheiro:
    ``OR`` (qualquer uma) por omissão, ``AND`` (todas) para segmentar a
    sério ("Sub 35" **e** "VIP").
    """
    limpas = normalizar_etiquetas(etiquetas)
    if not limpas:
        return None
    operador = "$all" if str(logica or "").strip().upper() == "AND" else "$in"
    return {CAMPO: {operador: limpas}}


async def listar_etiquetas_em_uso(
    *,
    tenant_condition: Optional[dict] = None,
) -> list[str]:
    """Etiquetas que existem mesmo, dentro do âmbito do utilizador.

    Nunca levanta: sem catálogo o filtro fica sem sugestões (e o
    utilizador pode sempre escrever a etiqueta), mas a página abre.
    """
    filtro: dict = dict(tenant_condition) if tenant_condition else {}
    try:
        valores = await db.processes.distinct(CAMPO, filtro)
    except Exception as e:  # noqa: BLE001 — degradação graciosa deliberada
        logger.warning("Não foi possível ler o catálogo de etiquetas: %s", e)
        return []

    etiquetas: list[str] = []
    vistas: set[str] = set()
    for valor in valores or []:
        texto = str(valor).strip()
        if not texto:
            continue
        chave = texto.casefold()
        if chave in vistas:
            continue
        vistas.add(chave)
        etiquetas.append(texto)
    return sorted(etiquetas, key=str.casefold)


async def run_get_process_labels(user: dict) -> dict[str, Any]:
    """Handler de ``GET /processes/labels`` — o catálogo para o filtro.

    O âmbito é o do UTILIZADOR (a rede), não o da empresa activa: a
    empresa activa é uma vista, a rede é a fronteira de segurança. É a
    mesma condição que as listagens usam, vinda do mesmo ponto único.
    """
    tenant_condition = await build_tenant_condition(user)
    etiquetas = await listar_etiquetas_em_uso(tenant_condition=tenant_condition)
    return {"labels": etiquetas, "count": len(etiquetas)}

File: frontend/src/utils/processLabels.js
/**
 * processLabels — etiquetas (tags) dos processos.
 *
 * O PEDIDO (Lote 5, Secção B, ponto 15)
 *   Marcadores personalizados ("Sub 35", "VIP", "Urgente") para
 *   segmentar processos, porque as fases já não chegam.
 *
 * O QUE JÁ EXISTIA
 *   `labels: List[str]` está no modelo do processo desde sempre e o
 *   `ProcessDetails` até renderiza os crachás. O que faltava era o
 *   EDITOR: o comentário no código dizia que a edição tinha sido
 *   "movida para um Dialog accionado pelo botão +" — esse Dialog nunca
 *   chegou a ser construído. O PACOTE DD removeu o cartão de Etiquetas
 *   e o substituto ficou por fazer, o que deixou o campo só acessível
 *   pela API.
 *
 * PORQUE É QUE A COR DERIVA DO TEXTO
 *   Guardar a cor obrigaria a uma segunda colecção de definições de
 *   etiqueta, ou a mudar `labels` para objectos — migrando os dados e
 *   todas as projecções. Derivá-la dá a mesma garantia sem nada para
 *   manter: "VIP" é da mesma cor em todos os ecrãs porque é a mesma
 *   palavra, não porque alguém a configurou igual em dois sítios.
 *
 * A NORMALIZAÇÃO ESPELHA A DO BACKEND (`services/process_labels.py`).
 * Divergirem daria um editor que aceita o que a API recusa.
 *
 * @module utils/processLabels
 */

/** Uma etiqueta é um marcador, não um campo de notas. */
export const MAX_COMPRIMENTO = 40;
/** Acima disto deixa de segmentar e passa a ser ruído no cabeçalho. */
export const MAX_ETIQUETAS = 20;

/**
 * Paleta de crachás. Tokens semânticos do Shadcn — a regra ESLint do
 * PACOTE 11 bloqueia cores Tailwind cruas, e um crachá com
 * `bg-blue-500` fica ilegível em dark mode.
 */
const PALETA = [
  "bg-primary/10 text-primary border-primary/20",
  "bg-secondary text-secondary-foreground border-border",
  "bg-accent text-accent-foreground border-border",
  "bg-muted text-muted-foreground border-border",
  "bg-destructive/10 text-destructive border-destructive/20",
];

const chave = (valor) => String(valor ?? "").trim().toLowerCase();

/**
 * Etiquetas limpas, sem repetições e em número comportável.
 *
 * Compara sem olhar à capitalização mas preserva a forma da primeira
 * ocorrência. Devolve sempre uma lista — vazia é uma resposta legítima
 * (limpar as etiquetas), não um erro.
 *
 * @param {string|Array} [valor]
 * @returns {string[]}
 */
export function normalizarEtiquetas(valor) {
  let cruas;
  if (typeof valor === "string") cruas = valor.split(",");
  else if (Array.isArray(valor)) cruas = valor;
  else return [];

  const etiquetas = [];
  const vistas = new Set();
  for (const crua of cruas) {
    if (crua == null) continue;
    const texto = String(crua).trim().slice(0, MAX_COMPRIMENTO).trim();
    if (!texto) continue;
    const k = chave(texto);
    if (vistas.has(k)) continue;
    vistas.add(k);
    etiquetas.push(texto);
    if (etiquetas.length >= MAX_ETIQUETAS) break;
  }
  return etiquetas;
}

/**
 * Classes do crachá de uma etiqueta, derivadas do próprio texto.
 *
 * Usa a chave normalizada: "VIP" e " vip " são a MESMA etiqueta depois
 * de normalizadas e não podem aparecer de cores diferentes.
 *
 * @param {string} [etiqueta]
 * @returns {string}
 */
export function corDaEtiqueta(etiqueta) {
  const k = chave(etiqueta);
  let hash = 0;
  for (let i = 0; i < k.length; i += 1) {
    hash = (hash * 31 + k.charCodeAt(i)) % 100000;
  }
  return PALETA[hash % PALETA.length];
}

/**
 * Se uma etiqueta pode ser acrescentada, e porquê não.
 *
 * O motivo não é decorativo: um botão que não faz nada e não diz porquê
 * é o silêncio do Bug 1 outra vez, noutro sítio.
 *
 * @param {string[]} existentes
 * @param {string} candidata
 * @returns {{ok: true, etiqueta: string} | {ok: false, motivo: string}}
 */
export function podeAcrescentar(existentes, candidata) {
  const actuais = normalizarEtiquetas(existentes);
  const [etiqueta] = normalizarEtiquetas([candidata]);

  if (!etiqueta) return { ok: false, motivo: "Escreva o nome da etiqueta." };
  if (actuais.some((e) => chave(e) === chave(etiqueta))) {
    return { ok: false, motivo: `A etiqueta "${etiqueta}" já está neste processo.` };
  }
  if (actuais.length >= MAX_ETIQUETAS) {
    return { ok: false, motivo: `Máximo de ${MAX_ETIQUETAS} etiquetas por processo.` };
  }
  return { ok: true, etiqueta };
}
